### ISSUE TITLE
Gloo Edge does not populate the identity object with the relevant data when wrapAsApiGateway is enabled

### ISSUE THREAD

**Issue by sadieleob** [solo-io/gloo#9541]
### Gloo Edge Product

Enterprise

### Gloo Edge Version

1.16.8

### Kubernetes Version

v1.27.13-eks-3af4770

### Describe the bug

When setting GE to mimic the AWS API GW behavior by setting to true **unwrapAsApiGateway** and **wrapAsApiGateway**, it looks like GE might not be wrapping the request in a way that Lambda function expects.

AWS Lambda logs:
`java.lang.NullPointerException: Cannot invoke "com.amazonaws.serverless.proxy.model.ApiGatewayRequestIdentity.getAccessKey()" because the return value of "com.amazonaws.serverless.proxy.model.AwsProxyRequestContext.getIdentity()" is null`

      

### Expected Behavior

GE wraps the request to AWS Lambda in a manner identical to how API Gateway does, including fields such as the identity in the event object.

### Steps to reproduce the bug

1. **Create AWS lambda**:

```
mvn archetype:generate -DgroupId=my.service -DartifactId=my-service -Dversion=1.0-SNAPSHOT \
       -DarchetypeGroupId=com.amazonaws.serverless.archetypes \
       -DarchetypeArtifactId=aws-serverless-jersey-archetype \
       -DarchetypeVersion=2.0.1
```
Run "mvn clean package" to generate a zip file.
Deploy the zip file in a Lambda function. Select Lambda Java version for the Lambda runtime same as what you have in pom.xml.

**AWS Lambda** 
- Runtime: Java 21, 
- Handler: my.service.StreamLambdaHandler::handleRequest,
- Arch: x86_64

**Testing**
Use following test event:
```
{
       "resource": "/ping",
       "path": "/ping",
       "httpMethod": "GET",
       "requestContext": {
           "stage": "stage1",
           "identity": {
             "apikey": "XXXXXXXXXXXXX",
             "userAgent": "curl/7.79.1"
           },
          "resourcePath": "/",
           "httpMethod": "GET",
           "path": "/"
       },
       "headers": {
            "Host": "www.example.com"
        },
        "multiValueHeaders": {
              "Host": ["www.example.com"]
        },
       "queryStringParameters": null,
       "multiValueQueryStringParameters": null,
       "pathParameters": null,
       "stageVariables": null,
       "body": null,
       "isBase64Encoded": false
   }
   ```

2. Define the following VS and US:
```
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: hello-lambda
  namespace: suresh
spec:
  aws:
    destinationOverrides:
      requestTransformation: false
      unwrapAsApiGateway: true
      wrapAsApiGateway: true
    lambdaFunctions:
    - logicalName: sadiel-mvn
      qualifier: $LATEST
    region: us-west-2
    secretRef:
      name: aws-creds
      namespace: gloo-system
  connectionConfig:
    maxRequestsPerConnection: 1
---                                                                                                 
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: cbp-portal
  namespace: default
spec:
  virtualHost:
    domains:
    - soloist.servebeer.com
    routes:
    - matchers:
      - prefix: /
      routeAction:
        single:
          destinationSpec:
            aws:
              logicalName: sadiel-mvn
          upstream:
            name: hello-lambda
            namespace: suresh
```

3. GE Settings
```
gloo:
    awsOptions:
      fallbackToFirstFunction: true
```

4. Requests:
```
curl -v -H"Host: soloist.servebeer.com" http://soloist.servebeer.com/v1/ping

curl -v -H"Host: soloist.servebeer.com" http://soloist.servebeer.com/v1/ping -H "api-key: $API_KEY" -H "Content-Type: application/json" -d '{"resource": "/ping","path": "/ping","httpMethod": "GET","requestContext": {"stage": "stage1","identity": {"apikey": "xxxxxxxxxx","userAgent": "curl/7.79.1"},"resourcePath": "/","httpMethod": "GET","path": "/"},"headers": {"Host": "www.example.com"},"multiValueHeaders": {"Host": ["www.example.com"]},"queryStringParameters": null,"multiValueQueryStringParameters": null,"pathParameters": null,"stageVariables": null,"body": null,"isBase64Encoded": false}'
```
5. Response:
{
  "upstream_host": "\"18.246.196.137:443\"",
  "method": "\"POST",
  "protocol": "HTTP/1.1\"",
  "path": "/2015-03-31/functions/sadiel-mvn/invocations?Qualifier=%24LATEST",
  "duration": 19,
  "response_code": 502,
  "upstreamCluster": "hello-lambda_suresh",
  "x-forwarded-for": "\"-\"",
  "response_flag": "-",
  "starttime": "[2024-05-29T17:28:57.019Z]",
  "request-id": "\"2699f1e0-6f5f-4947-9c70-65e7c6b48b8b\"",
  "requestHeaders": null,
  "upstreamHost": "18.246.196.137:443",
  "upstream_svc_time": null
}

**Comment by sadieleob** [comment]
tested in 1.16.10 and the issue persists. 

```
helm list -n gloo-system                                                                                                                               
NAME	NAMESPACE  	REVISION	UPDATED                             	STATUS  	CHART          	APP VERSION
gloo	gloo-system	17      	2024-06-17 09:54:43.728908 -0500 CDT	deployed	gloo-ee-1.16.10

curl -v -H"Host: soloist.servebeer.com" http://soloist.servebeer.com/v1/ping -H "api-key: $API_KEY" -H "Content-Type: application/json" -d '{"resource": "/ping","path": "/ping","httpMethod": "GET","requestContext": {"stage": "stage1","identity": {"apikey": "xxxxxxxxxxxxx","userAgent": "curl/7.79.1"},"resourcePath": "/","httpMethod": "GET","path": "/"},"headers": {"Host": "www.example.com"},"multiValueHeaders": {"Host": ["www.example.com"]},"queryStringParameters": null,"multiValueQueryStringParameters": null,"pathParameters": null,"stageVariables": null,"body": null,"isBase64Encoded": false}'

* Host soloist.servebeer.com:80 was resolved.
* IPv6: (none)
* IPv4: 35.167.216.82
*   Trying 35.167.216.82:80...
* Connected to soloist.servebeer.com (35.167.216.82) port 80
> POST /v1/ping HTTP/1.1
> Host: soloist.servebeer.com
> User-Agent: curl/8.6.0
> Accept: */*
> api-key: xxxxxxxxxx
> Content-Type: application/json
> Content-Length: 485
>
< HTTP/1.1 502 Bad Gateway
< content-type: application/json
< content-length: 29
< x-ratelimit-limit: 1000
< x-ratelimit-remaining: 999
< x-ratelimit-reset: 1
< date: Mon, 17 Jun 2024 15:06:28 GMT
< server: envoy
<
* Connection #0 to host soloist.servebeer.com left intact
{"message": "Gateway Timeout"}%

2024-06-17T14:51:45.840Z
java.lang.NullPointerException: Cannot invoke "com.amazonaws.serverless.proxy.model.ApiGatewayRequestIdentity.getAccessKey()" because the return value of "com.amazonaws.serverless.proxy.model.AwsProxyRequestContext.getIdentity()" is null

```

**Comment by DuncanDoyle** [comment]
We need to make sure that we compare the full requests that are being sent by Gloo Edge with the full requests that are being sent by the AWS API Gateway to verify that there are no other fields and objects not being populated.

**Comment by nrjpoddar** [comment]
@nfuden to look at this issue for scope/estimation.

**Comment by spunuru** [comment]
The behavior of **multiValueHeaders** and **multiValueQueryStringParameters** in the request object (passed by Gloo to Lambda upstream) is not compatible with how AWS API Gateway sends them.  Please see https://support.solo.io/hc/en-us/requests/4378 for additional details.

Appreciate if you can fix this issue as part of the issue.

**Comment by DuncanDoyle** [comment]
Additional analysis in this Google Doc: https://docs.google.com/document/d/1HYZ2TvqxKCyzYrCCOFyxXnu_OW4igP7M8Fu0966Yv_Y/edit?usp=sharing

The TL;DR here is that:

- AWS ApiGateway always populates a `/requestContext/identity` object, regardless of whether an identity has been passed, and some AWS parser libraries expect that `identity` object to be present. If it's not, this can result in, for example, NullPointerExceptions. So Gloo Gateway needs to generate it with the information available and pass it to the upstream lambda.
- AWS ApiGateway will set `/requestContext/identity/apiKey` to the value of the `x-api-key` (!!!!! not `api-key`) header if passed on the original request. Gloo Gateway needs to do the same.
- AWS ApiGateway always sets all headers in `multiValueHeaders` map, even if these headers are not multi-value. A lot of Amazon AWS library's parsers rely on this behaviour. See also https://aws.amazon.com/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway/ and https://aws.amazon.com/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway/. Gloo Gateway needs to have the same behaviour (more details in the Google Doc).
- AWS ApiGateway always sets all query parameters in the `multiValueQueryStringParameter` map, even if the parameters are not multi-value. Again, a lot of Amazon AWS library's parsers rely on this behaviour, so Gloo Gateway needs to have the same semantics.

Google Doc contains various example requests from AWS ApiGateway, comparing them with requests that Gloo Gateway generates.

**Comment by nfuden** [comment]
1.18.0-beta3

**Comment by spunuru** [comment]
Below requirement did not work for me. Can you please check?

AWS ApiGateway will set /requestContext/identity/apiKey to the value of the x-api-key (!!!!! not api-key) header if passed on the original request. Gloo Gateway needs to do the same.

**Comment by sadieleob** [comment]
feedback from customer:
```
According to [Input format of a Lambda function for proxy integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format), when a request "/foo?a=123&b=456&a=789", the "queryStringParameters" key should looks like this:

"queryStringParameters": {
  "a": "123,789",
  "b": "456"
}
However, you produced following data (this is the existing behavior before this beta):

"queryStringParameters": {
  "a": "789",
  "b": "456"
}
I do believe making this change to be the same as API gateway is likely not backwards compatible. Please investigate this carefully and proceed with caution.

At a minimal, please officially document this as "limitation" when you decide NOT to fix it.
```
